### DISCUSSION TITLE
Add option to skip getLabels for index.docker.io to reduce impact on Docker Hub rate limit

### DISCUSSION THREAD

**Discussion by raphaelzoellner** [renovatebot/renovate#29582] [Suggest an Idea]
### Tell us more.

DockerHub defines a [rate limit](https://docs.docker.com/docker-hub/download-rate-limit/) on GET `/v2/*/manifests/*` requests.

Renovate uses such requests to identify gitRef, sourceUrl and homepage for the release.
https://github.com/renovatebot/renovate/blob/7516af6e6aface68ca3b1d605162ff4e9af2b928/lib/modules/datasource/docker/index.ts#L1098-L1115

The labels seem to be already considered optional for docker library images.
https://github.com/renovatebot/renovate/blob/37.401.1/lib/modules/datasource/docker/index.ts#L459-L465

I suggest to add a self-hosted experimental environment variable `RENOVATE_X_DISABLE_DOCKER_HUB_LABELS` which disables all label requests to https://index.docker.io to optinally reduce the impact on the DockerHub rate limit further. 

https://docs.renovatebot.com/self-hosted-experimental/#self-hosted-experimental-environment-variables

related to: 
#17702
#19036
#21964

**Comment by rarkins** [comment]
I recommend that you set https://docs.renovatebot.com/self-hosted-experimental/#renovate_x_docker_hub_tags in order to reduce the limit from index.docker.io. It's still got rate limits but we don't seem to hit them as much due to its more efficient API results and it will be the default for Renovate v38

You're welcome to submit a PR for your idea too though, and report back on the findings. If the only queries which affect rate limits are for the URLs and we cache them heavily then I wouldn't expect it to use up much API calls anyway.

**Comment by raphaelzoellner** [comment]
I can confirm that caching keeps these requests at a minimum. So far working with Renovate is great! Thanks a lot for this great work!

In scenarios where the label information might not be needed at all (e.g. frequent dry-run mode lookups) especially when multiple independent systems that are running Renovate share the same IP towards DockerHub, it would be nice to opt out of the rate limited requests.

I was able to confirm that the label requests mentioned above were counted against the DockerHub [rate limit](https://docs.docker.com/docker-hub/download-rate-limit/).

I've added the proposed variable and tested the changes of #29624 on a test repository containing the following Dockerfile.

```yaml
FROM sonarsource/sonar-scanner-cli:4.8.0
```

I've tested the changes by running Renovate against the repository in dry-run=lookup mode.

Without the variable `RENOVATE_X_DISABLE_DOCKER_HUB_LABELS` I observed the following debug logs
```
DEBUG: getLabels(https://index.docker.io, sonarsource/sonar-scanner-cli, latest) (repository=local)
DEBUG: getManifestResponse(https://index.docker.io, sonarsource/sonar-scanner-cli, latest, get) (repository=local)
```
After querying the DockerHub Rate Limit I noticed a decrease in remaining.

After setting the variable `RENOVATE_X_DISABLE_DOCKER_HUB_LABELS` I observed the added debug logs in the PR.
```
DEBUG: getLabels(https://index.docker.io, sonarsource/sonar-scanner-cli, latest) (repository=local)
DEBUG: Docker Hub image - skipping label lookup due to RENOVATE_X_DISABLE_DOCKER_HUB_LABELS (repository=local)
```
After querying the DockerHub Rate Limit time I noticed no decrease.

**Reply by rarkins** [reply]
Does this change mean that Renovate itself doesn't impact the rate limit? So for example you could use Renovate without any token and not hit limits?

**Reply by viceice** [reply]
if you check the docker hub that frequently, then you should have a local proxy, so the manifest is cached on the proxy.

 it's really easy with official registry docker image and renovate registry aliases.

**Reply by raphaelzoellner** [reply]
> Does this change mean that Renovate itself doesn't impact the rate limit? So for example you could use Renovate without any token and not hit limits?

I've had to take a deeper look into the callstack to answer this. The `GET /v2/*/manifests/*` requests originate only from `getManifestResponse` in the docker datasource.

### The short answer

No, when digests are used or should be pinned there might be occur the mentioned calls even though the variable is specified.

### The detailed answer

I followed the calls made to `getManifestResponse`.

- [getManifestResponse](https://github.com/renovatebot/renovate/blob/37.401.1/lib/modules/datasource/docker/index.ts#L96)(mode: get)
    - [getManifest](https://github.com/renovatebot/renovate/blob/37.401.1/lib/modules/datasource/docker/index.ts#L279)
        - [getConfigDigest](https://github.com/renovatebot/renovate/blob/37.401.1/lib/modules/datasource/docker/index.ts#L268)
            - [getImageArchitecture](https://github.com/renovatebot/renovate/blob/37.401.1/lib/modules/datasource/docker/index.ts#L354)
                - [getDigest](https://github.com/renovatebot/renovate/blob/37.401.1/lib/modules/datasource/docker/index.ts#L805)
                    - [getDigest](https://github.com/renovatebot/renovate/blob/37.401.1/lib/modules/datasource/docker/index.ts#L805)                        
        - [getManifest](https://github.com/renovatebot/renovate/blob/37.401.1/lib/modules/datasource/docker/index.ts#L279) (first manifest in case of manifest lists)
        - [getLabels](https://github.com/renovatebot/renovate/blob/37.401.1/lib/modules/datasource/docker/index.ts#L452)
            - [getReleases](https://github.com/renovatebot/renovate/blob/37.401.1/lib/modules/datasource/docker/index.ts#L1041C9-L1041C20)

So there seem two paths inside the docker datasource end at `getDigest` and `getReleases` which implement/override the abstract Datasource class.

- [getDigest](https://github.com/renovatebot/renovate/blob/37.401.1/lib/modules/datasource/datasource.ts#L41)
    - [lookupUpdates](https://github.com/renovatebot/renovate/blob/37.401.1/lib/workers/repository/process/lookup/index.ts#L50)
        if (config.pinDigests === true || config.currentDigest): when digest is specified or when digest should be pinned via config

- [getReleases](https://github.com/renovatebot/renovate/blob/37.401.1/lib/modules/datasource/datasource.ts#L37)
    - [getRegistryReleases](https://github.com/renovatebot/renovate/blob/37.401.1/lib/modules/datasource/index.ts#L60)
        - [firstRegistry](https://github.com/renovatebot/renovate/blob/37.401.1/lib/modules/datasource/index.ts#L90)
            - [fetchReleases](https://github.com/renovatebot/renovate/blob/37.401.1/lib/modules/datasource/index.ts#L254C16-L254C29)
        - [huntRegistries](https://github.com/renovatebot/renovate/blob/37.401.1/lib/modules/datasource/index.ts#L109C16-L109C30)
            - [fetchReleases](https://github.com/renovatebot/renovate/blob/37.401.1/lib/modules/datasource/index.ts#L254C16-L254C29)
        - [mergeRegistries](https://github.com/renovatebot/renovate/blob/37.401.1/lib/modules/datasource/index.ts#L140)
            - [fetchReleases](https://github.com/renovatebot/renovate/blob/37.401.1/lib/modules/datasource/index.ts#L254C16-L254C29)
    - [fetchReleases](https://github.com/renovatebot/renovate/blob/37.401.1/lib/modules/datasource/index.ts#L254)

`fetchReleases` is then used by `getPkgReleases`

- [fetchReleases](https://github.com/renovatebot/renovate/blob/37.401.1/lib/modules/datasource/index.ts#L254C16-L254C29)
    - [fetchCachedReleases](https://github.com/renovatebot/renovate/blob/37.401.1/lib/modules/datasource/index.ts#L315C10-L315C29)
        - [getRawPkgReleases](https://github.com/renovatebot/renovate/blob/37.401.1/lib/modules/datasource/index.ts#L333)
            - [getPkgReleases](https://github.com/renovatebot/renovate/blob/37.401.1/lib/modules/datasource/index.ts#L378)
            - [lookupUpdates](https://github.com/renovatebot/renovate/blob/37.401.1/lib/workers/repository/process/lookup/index.ts#L50)

`getPkgReleases` is used in a number of places e.g. [managers](https://github.com/renovatebot/renovate/tree/37.401.1/lib/modules/manager), but also [changelogs](https://github.com/renovatebot/renovate/blob/37.401.1/lib/workers/repository/update/pr/changelog/releases.ts)

> if you check the docker hub that frequently, then you should have a local proxy, so the manifest is cached on the proxy.
> 
> it's really easy with official registry docker image and renovate registry aliases.

Do you mean a proxy OCI registry? Wouldn't that lead to Renovate only seeing the tags in the proxy registry, which might not correspond to the updated tags until they have been pulled through the proxy OCI registry?

For use cases when the actual pulled image, but not the information about updates shall be fetched from the registry we already have such a proxy OCI registry in place.

But for use cases like renovate dry-run mode used to look up only potential updates or testing your local renovate.json configuration in the repository I think the information gained from the labels might not always be needed. Especially when they can only be gained via manifest requests `GET /v2/*/manifests/*` to `https://index.docker.io` that seem to have a stricter rate limit than requests to `https://hub.docker.com`.

**Reply by viceice** [reply]
no, most docker proxies will pass through the tags list ( not counted for rate limit). I'm using the official registry docker image as pull through proxy without any issues while running renovate every half hour

**Reply by raphaelzoellner** [reply]
I tried specifying registry aliases, but wasn't able to get it working when the registry is not explicitly defined in the Dockerfile like in the example above.

```json
  "registryAliases": {
    "index.docker.io": "harbor.example.com/dockerhub",
    "docker.io": "harbor.example.com/dockerhub",
    "registry-1.docker.io": "harbor.example.com/dockerhub"
  }
```

```yaml
FROM sonarsource/sonar-scanner-cli:4.8.0
```

I suspect the reason for this is that `fromImage` is not prefixed with a default registry if no registry is specified, usually docker hub is used or another default registry mirror specified for to the image build tool (docker, kaniko, ...).
https://github.com/renovatebot/renovate/blob/37.401.1/lib/modules/manager/dockerfile/extract.ts#L335-L370

If this is the case then the regex used to match registryAliases will not match and therefore no replacement is taking place.
https://github.com/renovatebot/renovate/blob/37.401.1/lib/modules/manager/dockerfile/extract.ts#L178-L181

On the other hand, this was working when the docker hub registry was explicitly set in the Dockerfile.

```yaml
FROM docker.io/sonarsource/sonar-scanner-cli:4.8.0
```

Another thing I noticed is that this was also working for Harbor, so I assume the Harbor logic changed regarding tags such that tag list requests are also passed through to docker hub.
https://github.com/renovatebot/renovate/discussions/9004#discussioncomment-3603677

Sadly we cannot assume that all scanned Dockerfiles will have a set registry. 
It might be worth introducing the ability to define default docker registries for Renovate, but I think this is out of scope for this discussion.

**Reply by viceice** [reply]
you can use registry url to override default docker hub .

**Reply by raphaelzoellner** [reply]
Thanks a lot, that's exactly what I was looking for. I completely missed this option while looking through the docs.

We will add this to our onboarding configuration.
